@DataSourceDefinition defined data source can't be used in persistence.xml

Defining a data source from within the application using either @DataSourceDefinition on a class or the data-source element in web.xml, and then using this in persistence.xml will cause a deployment failure.

See https://java.net/jira/browse/GLASSFISH-20944

@smillidge 

A possible related exception was thrown here: https://travis-ci.org/javaee-samples/javaee7-samples/jobs/93063766

This test case is similar to the one I mentioned above, but it used the XA data source:

```
<data-source>
     <name>java:app/MyApp/MyDS</name>
     <!-- org.hsqldb.jdbc.JDBCDataSource is non-XA alternative -->
     <class-name>org.hsqldb.jdbc.pool.JDBCXADataSource</class-name>
     <url>jdbc:hsqldb:mem:realtime;user=test;password=testx</url>
 </data-source>
```

In this case it throws:

```
Caused by: java.sql.SQLInvalidAuthorizationSpecException: invalid authorization specification
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.pool.JDBCXADataSource.getXAConnection(Unknown Source)
	at com.sun.gjc.spi.XAManagedConnectionFactory.createManagedConnection(XAManagedConnectionFactory.java:118)
	... 144 more
Caused by: org.hsqldb.HsqlException: invalid authorization specification
	at org.hsqldb.error.Error.error(Unknown Source)
	at org.hsqldb.error.Error.error(Unknown Source)
	... 146 more
```

For the other containers hsqldb doesn't throw this exception, so I assume it's related to how GlassFish sets the username/password.

--- a/appserver/persistence/jpa-container/src/main/java/org/glassfish/persistence/jpa/PersistenceUnitLoader.java
+++ b/appserver/persistence/jpa-container/src/main/java/org/glassfish/persistence/jpa/PersistenceUnitLoader.java
@@ -104,8 +104,15 @@ public class PersistenceUnitLoader {
        // This should be removed once version of EclipseLink which fixes the issue is integrated.
        // set the system property required by EclipseLink before we load it.
        setSystemPropertyToEnableDoPrivilegedInEclipseLink();
-
+       
+        //NOTE If we are going to create datasources we must switch to the final classloader otherwise 
+        // the datasource will have the wrong classloader associated with the classes.
+        // This is to fix PAYARA-547
+        ClassLoader cl = Thread.currentThread().getContextClassLoader();
+        Thread.currentThread().setContextClassLoader(providerContainerContractInfo.getClassLoader());
+ 
        emf = loadPU(puToInstatntiate);
+       Thread.currentThread().setContextClassLoader(cl);
    }
 
     /**

